multi-tenancy-observability/README.md

```diff
@@ -0,0 +1,270 @@
+
+We are considering a number of options that are not ideal at the moment:
+
+- having double datasource: one for customer with OAuth authentication and another dedicated to GiantSwarm people authenticated throw teleport.
+- having double grafana (same idea than above but for the all grafana instance, not just the datasource).
+- share an authentication secret via cookies.
+- other ideas are welcome!
+
+#### Mapping user's groups to tenants
+
+What interface could we offer our customers to enable them to define the mapping between their groups and their tenants:
+
+- Custom resources like TenantCR, GroupCR
+- UI in happa
+- other ideas ?
+
+## Write path
+
+### Current architecture for the write path
+
+On the write path, we currently use promtail to collect container and machine logs and grafana agent to collect kubernetes events.
+
+We currently need those 2 tools because:
+1. Promtail cannot get kubernetes events
+2. Grafana Agent is going through quite a big rewrite in it's configuration and all features we needed were not supported when we deployed it
+
+Here is what the flow looks like:
+1. The logging-operator creates a `password` for the cluster's logging agents to be able to send logs to Loki and configures them (actual configuration of what to scrape and so on) via an extra-config.
+2. The logging-operator adds the new cluster_id/password pair to the proxy configuration
+3. The logging agents collect and send logs to the loki-multi-tenant-proxy which validates the basic authentication sent in the http header and transform it into a tenant information (X-Scope_OrgID)
+4. Loki writes the data to storage for the configured tenant (cluster_id in that case)
+
+### Configuring multi-tenancy on the write path
+
+#### Different tenant configuration options
+
+After a few different customer calls, we found out that they have different needs when it comes to multi-tenancy.
+
+##### Cluster == Tenant
+
+This is the approach we are currently using for technical reasons. We do not think this would match with customers workload as they mostly have a lot of customers sharing clusters.
+
+##### Namespace == Tenant
+
+This approach makes it really easy for us and customers to configure tenants and it would definitely be easy to onboard teams (e.g. adding a label on the namespace would make the agents collect logs from the containers in the namespace).
+
+But this approach brings a few questions:
+1. What tenancy should common internal apps use (e.g. prometheus-operator, api-server, cilium logs are accessed by multiple internal teams)? Do we need to ensure customers are putting those apps into a namespace accessible by all?
+2. What about teams that share namespaces?
+
+##### App == Tenant
+
+This approach is especially flexible for customers but we do not think this approach makes sense because the mapping configuration would be impossible to maintain and we would probably face technical limits pretty fast
```

Comment:
IMO this is the natural end state.

Two valid requests:
- "The team that owns namespace X can view logs from any application in that namespace"
- "The monitoring team can view prometheus logs from any namespace in any cluster where they run prometheus"

Roughly app-level granularity is inevitable in my mind. It also works well with other potential features (described in the general comment).

Reply:
It could be but then I do not think it works for all things.

Usually, customers run applications for one team in one namespace with a clear boundary. Sure, prometheus might be a bit different but that's really dependant on the customers organizations.

What I do not like in this is that if you want all the logs of all apps in one namespace, then loki will actually run at least number of apps queries. Sure they're paralellized and so on but that does look a bit too much to me for most use cases.

Reply:
> loki will actually run at least number of apps queries

The actual query implementation is beyond my knowledge :sweat_smile: Sure, it might be prohibitively expensive or that this is intentionally nerfed in OSS and reserved for enterprise versions.

BUT to the greatest extent possible, I would double down on the likelihood of needing app-level granularity support. Some other examples:
- any type of less-trusted collaborators (externals, students, etc.) who should have access to only certain components inside a namespace
- selective access to "system" resources - for example, there are multiple types of users who will need access to coredns or nginx, but they shouldn't have access to everything in the namespace, especially kube-system stuff. 
- automation - similar to above, but for machine accounts. 

If it is possible to change the tenant behavior later, maybe logging-operator could default to tenant-per-ns until the first app-level rule is applied, or create app tenants only for those targeted by allow rules. This probably means the capability would only apply to future logs, which is a tradeoff, but could be a workable middle ground.

Reply:
Yes, from what I see, we can add levels of granularity on-demand, and we could even do that opt-in for apps, so it's not by default tenanted by app, but there's additional tenants for apps where we see demand for selective access control (e.g. we might at some point pull nginx, or coredns tenants into our default mapping). WDYT?

Reply:
Yes, from what I understood a new tenant is a new and separate datastore, which makes duplication of data a topic in that context, but also the issues you mentioned, you basically won't be able to access historical data previous to tenant creation. Not sure about how that looks with queries, but that I would assume should not change as tenants should not influence data, but you never know, maybe @QuentinBisson knows more.

Reply:
Yes you can think of it as a datastore and a tenant is only a http header so there is no reason why customers that want to migrate to 1 tenant per app would lose old logs if they keep the old and new tenants together :)